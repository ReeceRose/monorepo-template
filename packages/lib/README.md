# Lib package

Common TypeScript code shared across TypeScript apps. Can be expanded to support other languages (like Go via modules).