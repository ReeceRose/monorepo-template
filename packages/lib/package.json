{
  "name": "lib",
  "version": "0.1.0",
  "scripts": {
    "test": "echo \"Error: no test specified\""
  },
  "devDependencies": {
    "@types/node": "^17.0.16",
    "@typescript-eslint/eslint-plugin": "^5.11.0",
    "@typescript-eslint/parser": "^5.11.0",
    "config": "workspace:*",
    "eslint": "^8.8.0",
    "eslint-config-prettier": "^8.3.0",
    "eslint-plugin-prettier": "^4.0.0",
    "tsconfig": "workspace:*",
    "typescript": "^4.5.5"
  }
}
