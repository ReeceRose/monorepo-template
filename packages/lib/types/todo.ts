export interface ITodo {
  uuid: string;
  description: string;
  completed: boolean;
}
