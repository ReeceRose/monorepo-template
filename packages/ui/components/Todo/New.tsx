import { useState } from 'react';

type Props = {
  insertTodo: (description: string) => Promise<void>;
};

export const NewTodo = ({ insertTodo }: Props): JSX.Element => {
  const [description, setDescription] = useState('');

  return (
    <form
      onSubmit={(e) => {
        e.preventDefault();
        insertTodo(description);
      }}
    >
      <div className="mb-6">
        <label
          htmlFor="description"
          className="block mb-2 text-sm font-medium text-gray-900 dark:text-gray-300"
        >
          Description
        </label>
        <input
          type="text"
          id="description"
          value={description}
          onChange={(e) => setDescription(e.target.value)}
          className="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-700 focus:border-blue-700 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-700 dark:focus:border-blue-700"
          required
        />
      </div>
      <button
        type="submit"
        className="text-white bg-blue-500 hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 font-medium rounded-lg text-sm w-full sm:w-auto px-5 py-2.5 text-center dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800"
      >
        Insert
      </button>
    </form>
  );
};
