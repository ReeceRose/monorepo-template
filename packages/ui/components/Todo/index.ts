export * from './Item';
export * from './List';
export * from './New';
