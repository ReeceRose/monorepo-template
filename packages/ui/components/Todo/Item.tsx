import { ITodo } from 'lib/types';

type Props = {
  data: ITodo;
  toggleCompleted: (id: string, completed: boolean) => Promise<void>;
};

export const TodoItem = ({ data, toggleCompleted }: Props): JSX.Element => {
  return (
    <tr className="bg-white border-b dark:bg-gray-800 dark:border-gray-700">
      <td className="px-6 py-4 text-sm text-gray-500 whitespace-nowrap dark:text-gray-400">
        {data.description}
      </td>
      <td className="px-6 py-4 text-sm text-gray-500 capitalize whitespace-nowrap dark:text-gray-400">
        {data.completed.toString()}
      </td>
      <td className="px-6 py-4 text-sm font-medium text-right whitespace-nowrap">
        <button
          onClick={(e) => {
            e.preventDefault();
            toggleCompleted(data.id, !data.completed);
          }}
          className="text-blue-600 dark:text-blue-500 hover:underline"
        >
          Toggle completed
        </button>
        <button
          onClick={(e) => {
            e.preventDefault();
            toggleCompleted(data.id, !data.completed);
          }}
          className="ml-4 text-red-600 hover:underline"
        >
          Delete
        </button>
      </td>
    </tr>
  );
};
