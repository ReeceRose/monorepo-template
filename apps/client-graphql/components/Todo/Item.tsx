import { useMutation } from 'urql';
import {
  DeleteTodoDocument,
  DeleteTodoMutation,
  UpdateTodoDocument,
  UpdateTodoMutation,
} from '@/generated/graphql';
import { ITodo } from 'lib/types';

import { TodoItem } from 'ui/components/Todo';
import { TodoItemProps } from 'lib/types/props/todo';

export const TodoItemWrapper = ({ data }: TodoItemProps): JSX.Element => {
  const [, sendUpdate] = useMutation<UpdateTodoMutation>(UpdateTodoDocument);
  const [, sendDelete] = useMutation<DeleteTodoMutation>(DeleteTodoDocument);

  const toggleCompleted = async (todo: ITodo): Promise<void> => {
    const { error } = await sendUpdate({
      todo: {
        id: todo.id,
        description: todo.description,
        completed: !todo.completed,
      },
    });
    if (error) {
      console.log('Failed to fetch and udpdate todo with GraphQL');
    }
  };

  const deleteTodo = async (id: string): Promise<void> => {
    const { error } = await sendDelete({
      id,
    });
    if (error) {
      console.log('Failed to delete todo from GraphQL');
    }
  };

  return (
    <TodoItem
      data={data}
      deleteTodo={deleteTodo}
      toggleCompleted={toggleCompleted}
    />
  );
};
