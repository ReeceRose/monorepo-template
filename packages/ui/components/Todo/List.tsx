import { ITodo } from 'lib/types';

import { TodoItem } from './Item';

type Props = {
  data?: ITodo[];
  toggleCompleted: (todo: ITodo) => Promise<void>;
  deleteTodo: (id: string) => Promise<void>;
};

export const TodoList = ({
  data,
  toggleCompleted,
  deleteTodo,
}: Props): JSX.Element => {
  return (
    <div className="flex flex-col mx-auto sm:w-full md:w-8/12">
      <div className="overflow-x-auto sm:-mx-6 lg:-mx-8">
        <div className="inline-block min-w-full py-2 sm:px-6 lg:px-8">
          <div className="overflow-hidden shadow-md sm:rounded-lg">
            <table className="min-w-full">
              <thead className="bg-gray-50 dark:bg-gray-700">
                <tr>
                  <th
                    scope="col"
                    className="w-2/3 px-6 py-3 text-xs font-medium tracking-wider text-left text-gray-700 uppercase dark:text-gray-400"
                  >
                    Description
                  </th>
                  <th
                    scope="col"
                    className="w-1/4 px-6 py-3 text-xs font-medium tracking-wider text-left text-gray-700 uppercase dark:text-gray-400"
                  >
                    Completed
                  </th>
                  <th scope="col" className="relative px-6 py-3">
                    <span className="sr-only">Actions</span>
                  </th>
                </tr>
              </thead>
              <tbody>
                {data && data.length != 0 ? (
                  data?.map((todo: ITodo) => (
                    <TodoItem
                      key={todo.id}
                      toggleCompleted={toggleCompleted}
                      deleteTodo={deleteTodo}
                      data={todo}
                    />
                  ))
                ) : (
                  <tr className="bg-white border-b dark:bg-gray-800 dark:border-gray-700">
                    <td className="px-6 py-4 text-sm font-medium text-gray-900 whitespace-nowrap dark:text-white">
                      No data
                    </td>
                  </tr>
                )}
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  );
};
