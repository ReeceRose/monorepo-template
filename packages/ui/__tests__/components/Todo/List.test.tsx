import { render } from '@testing-library/react';
import { ITodo } from 'lib/types';

import { TodoList, TodoItem as TodoItemUI } from '../../../components/Todo';

const todo: ITodo = {
  id: 'test',
  description: 'test',
  completed: false,
};

const todoItemWrapper = (): JSX.Element => {
  const deleteTodo = () => Promise.resolve();
  const toggleCompleted = () => Promise.resolve();
  return (
    <TodoItemUI
      data={todo}
      deleteTodo={deleteTodo}
      toggleCompleted={toggleCompleted}
    />
  );
};

test('TodoList component matches snapshot', () => {
  const { asFragment } = render(
    <TodoList TodoItem={todoItemWrapper} data={[todo]} />
  );
  expect(asFragment()).toMatchSnapshot();
});
