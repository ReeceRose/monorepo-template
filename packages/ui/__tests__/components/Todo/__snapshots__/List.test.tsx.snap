// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col"
  >
    <div
      class="overflow-x-auto sm:-mx-6 lg:-mx-8"
    >
      <div
        class="inline-block min-w-full py-2 sm:px-6 lg:px-8"
      >
        <div
          class="overflow-hidden shadow-md sm:rounded-lg"
        >
          <table
            class="min-w-full"
          >
            <thead
              class="bg-gray-50 dark:bg-gray-700"
            >
              <tr>
                <th
                  class="w-2/3 px-6 py-3 text-xs font-medium tracking-wider text-left text-gray-700 uppercase dark:text-gray-400"
                  scope="col"
                >
                  Description
                </th>
                <th
                  class="w-1/4 px-6 py-3 text-xs font-medium tracking-wider text-left text-gray-700 uppercase dark:text-gray-400"
                  scope="col"
                >
                  Completed
                </th>
                <th
                  class="relative px-6 py-3"
                  scope="col"
                >
                  Actions
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class="bg-white border-b dark:bg-gray-800 dark:border-gray-700"
              >
                <td
                  class="px-6 py-4 text-sm text-gray-500 whitespace-nowrap dark:text-gray-400"
                >
                  test
                </td>
                <td
                  class="px-6 py-4 text-sm text-gray-500 capitalize whitespace-nowrap dark:text-gray-400"
                >
                  false
                </td>
                <td
                  class="px-6 py-4 text-sm font-medium text-right whitespace-nowrap"
                >
                  <button
                    class="text-blue-600 dark:text-blue-500 hover:underline"
                  >
                    Toggle completed
                  </button>
                  <button
                    class="ml-4 text-red-600 hover:underline"
                  >
                    Delete
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
